File: backend/src/routes/uid.js
import { Router } from 'express';
import { query, one, tx } from '../db/pool.js';
import { asyncHandler } from '../middleware/error.js';
import { requireAuth, requireManager, requireSupervisor, requireOperator, HttpError } from '../middleware/auth.js';
import { writeAudit } from '../middleware/audit.js';
import { UID_SELECT, serializeUid, getUid } from '../utils/serializers.js';
import { bulkCreateUids, completeStep, doConverting, qcSignoff } from '../services/uidService.js';

const router = Router();

// ── Lookup (service team + all) ────────────────────────────────────────────
router.get(
  '/lookup/:code',
  requireAuth,
  asyncHandler(async (req, res) => {
    const data = await getUid({ code: req.params.code.toUpperCase(), includeHistory: true });
    if (!data) throw new HttpError(404, 'UID not found');
    res.json(data);
  })
);

// ── Operator queue ─────────────────────────────────────────────────────────
router.get(
  '/queue/operator',
  requireAuth,
  asyncHandler(async (req, res) => {
    const locId = req.query.location_id ? parseInt(req.query.location_id, 10) : req.user.primary_location_id;
    const params = [];
    let where = "u.status = 'active'";
    if (locId) {
      params.push(locId);
      where += ` AND u.factory_location_id = $${params.length}`;
    }
    const rows = await query(
      `${UID_SELECT} WHERE ${where} ORDER BY u.priority DESC, u.created_at LIMIT 200`,
      params
    );
    res.json(rows.map(serializeUid));
  })
);

// ── QC pending queue ─────────────────────────────────────────────────────────
// UIDs whose current step is a QC step and which are still active.
// Ordered before the /:id routes so /qc/pending is matched as a literal path.
router.get(
  '/qc/pending',
  requireAuth,
  asyncHandler(async (req, res) => {
    const params = [];
    let where = "u.status = 'active' AND cs.is_qc_step = TRUE";
    if (req.query.location_id) {
      params.push(parseInt(req.query.location_id, 10));
      where += ` AND u.factory_location_id = $${params.length}`;
    }
    const rows = await query(
      `${UID_SELECT} WHERE ${where} ORDER BY u.priority DESC, u.created_at LIMIT 200`,
      params
    );
    res.json(rows.map(serializeUid));
  })
);

// ── QC sign-off ──────────────────────────────────────────────────────────────
router.post(
  '/:uidId/qc-signoff',
  requireOperator,
  asyncHandler(async (req, res) => {
    const uidId = parseInt(req.params.uidId, 10);
    const b = req.body || {};
    if (!['pass', 'fail', 'borderline'].includes(b.result)) {
      throw new HttpError(400, "result must be 'pass', 'fail', or 'borderline'");
    }
    await qcSignoff({
      uidId,
      performedById: req.user.id,
      result: b.result,
      values: b.values ?? null,
      notes: b.notes ?? null,
      workstationId: b.workstation_id ?? null,
    });
    await writeAudit(req.user.id, 'qc-signoff', 'uids', uidId, null, { result: b.result });
    res.json(await getUid({ id: uidId }));
  })
);

// ── List / search ──────────────────────────────────────────────────────────
router.get(
  '/',
  requireAuth,
  asyncHandler(async (req, res) => {
    const { location_id, cycle_type_id, status, search } = req.query;
    const skip = parseInt(req.query.skip || '0', 10);
    const limit = parseInt(req.query.limit || '100', 10);

    const conds = [];
    const params = [];
    if (location_id) { params.push(parseInt(location_id, 10)); conds.push(`u.factory_location_id = $${params.length}`); }
    if (cycle_type_id) { params.push(parseInt(cycle_type_id, 10)); conds.push(`u.cycle_type_id = $${params.length}`); }
    if (status) { params.push(status); conds.push(`u.status = $${params.length}`); }
    if (search) { params.push(`%${search}%`); conds.push(`u.code ILIKE $${params.length}`); }
    const where = conds.length ? `WHERE ${conds.join(' AND ')}` : '';

    const totalRow = await one(`SELECT COUNT(*)::int AS n FROM uids u ${where}`, params);
    const rows = await query(
      `${UID_SELECT} ${where} ORDER BY u.id DESC OFFSET $${params.length + 1} LIMIT $${params.length + 2}`,
      [...params, skip, limit]
    );
    res.json({ total: totalRow.n, items: rows.map(serializeUid) });
  })
);

// ── Bulk create ────────────────────────────────────────────────────────────
router.post(
  '/bulk-create',
  requireManager,
  asyncHandler(async (req, res) => {
    const b = req.body || {};
    if (!b.quantity || b.quantity < 1 || b.quantity > 500) throw new HttpError(400, 'Quantity must be 1–500');
    const uids = await bulkCreateUids({
      quantity: b.quantity,
      cycleTypeId: b.cycle_type_id,
      factoryLocationId: b.factory_location_id,
      createdById: req.user.id,
      productTypeId: b.product_type_id ?? null,
      sizeId: b.size_id ?? null,
      designId: b.design_id ?? null,
      priority: b.priority ?? 'normal',
      moId: b.mo_id ?? null,
      receivingEventId: b.receiving_event_id ?? null,
    });
    await writeAudit(req.user.id, 'bulk-create', 'uids', null, null, { codes: uids.map((u) => u.code) });
    res.status(201).json({ created: uids.length, uids });
  })
);

// ── Bulk cycle change (before any steps) ───────────────────────────────────
router.post(
  '/bulk-change-cycle',
  requireManager,
  asyncHandler(async (req, res) => {
    const { uid_ids = [], new_cycle_type_id } = req.body || {};
    const updated = await tx(async (c) => {
      const newVersion = await c.one(
        'SELECT * FROM cycle_versions WHERE cycle_type_id = $1 AND is_current = TRUE',
        [new_cycle_type_id]
      );
      if (!newVersion) throw new HttpError(404, 'Cycle type or version not found');
      const out = [];
      for (const uidId of uid_ids) {
        const uid = await c.one('SELECT * FROM uids WHERE id = $1', [uidId]);
        if (!uid) continue;
        const hist = await c.one('SELECT 1 FROM uid_step_history WHERE uid_id = $1 LIMIT 1', [uidId]);
        if (hist) throw new HttpError(400, `UID ${uid.code} has steps completed — cannot change cycle`);
        await c.query('UPDATE uids SET cycle_type_id = $1, cycle_version_id = $2 WHERE id = $3', [
          new_cycle_type_id,
          newVersion.id,
          uidId,
        ]);
        out.push(uid.code);
      }
      return out;
    });
    res.json({ updated });
  })
);

// ── Step completion ────────────────────────────────────────────────────────
router.post(
  '/:uidId/complete-step',
  requireOperator,
  asyncHandler(async (req, res) => {
    const uidId = parseInt(req.params.uidId, 10);
    const b = req.body || {};
    await completeStep({
      uidId,
      performedById: req.user.id,
      workstationId: b.workstation_id,
      qcResult: b.qc_result ?? null,
      qcValues: b.qc_values ?? null,
      notes: b.notes ?? null,
    });
    res.json(await getUid({ id: uidId }));
  })
);

// ── Converting ─────────────────────────────────────────────────────────────
router.post(
  '/:uidId/convert',
  requireSupervisor,
  asyncHandler(async (req, res) => {
    const uidId = parseInt(req.params.uidId, 10);
    const b = req.body || {};
    const childIds = await doConverting({
      parentUidId: uidId,
      supervisorId: req.user.id,
      children: b.children || [],
      patternId: b.pattern_id ?? null,
    });
    const children = [];
    for (const id of childIds) children.push(await getUid({ id }));
    await writeAudit(req.user.id, 'convert', 'uids', uidId, null, { children: children.map((c) => c.code) });
    res.json({ parent_uid_id: uidId, children });
  })
);

// ── Design confirmation ────────────────────────────────────────────────────
router.post(
  '/:uidId/confirm-design',
  requireManager,
  asyncHandler(async (req, res) => {
    const uidId = parseInt(req.params.uidId, 10);
    const { design_id, size_id = null } = req.body || {};
    const uid = await one('SELECT * FROM uids WHERE id = $1', [uidId]);
    if (!uid) throw new HttpError(404, 'UID not found');
    if (uid.design_locked) throw new HttpError(400, 'Design is locked — cannot change after Step 17');

    let setSize = '';
    const params = [design_id];
    if (size_id) {
      const valid = await one(
        'SELECT 1 FROM design_size_validity WHERE design_id = $1 AND size_id = $2',
        [design_id, size_id]
      );
      if (!valid) throw new HttpError(400, 'Invalid design-size combination');
      params.push(size_id);
      setSize = `, size_id = $${params.length}`;
    }
    const statusReset = uid.status === 'on_hold' ? ", status = 'active'" : '';
    params.push(uidId);
    await query(
      `UPDATE uids SET design_id = $1, design_confirmed = TRUE${setSize}${statusReset} WHERE id = $${params.length}`,
      params
    );
    await writeAudit(req.user.id, 'confirm-design', 'uids', uidId, { design_id: uid.design_id }, { design_id });
    res.json(await getUid({ id: uidId }));
  })
);

// ── MO linking ─────────────────────────────────────────────────────────────
router.post(
  '/:uidId/link-mo/:moId',
  requireManager,
  asyncHandler(async (req, res) => {
    const uidId = parseInt(req.params.uidId, 10);
    const moId = parseInt(req.params.moId, 10);
    const uid = await one('SELECT id FROM uids WHERE id = $1', [uidId]);
    if (!uid) throw new HttpError(404, 'UID not found');
    await query('UPDATE uids SET mo_id = $1 WHERE id = $2', [moId, uidId]);
    res.json({ uid_id: uidId, mo_id: moId });
  })
);

// ── Inter-location transfer ────────────────────────────────────────────────
router.post(
  '/:uidId/transfer',
  requireSupervisor,
  asyncHandler(async (req, res) => {
    const uidId = parseInt(req.params.uidId, 10);
    const { to_location_id, reason } = req.body || {};
    await tx(async (c) => {
      const uid = await c.one('SELECT * FROM uids WHERE id = $1', [uidId]);
      if (!uid) throw new HttpError(404, 'UID not found');
      await c.query(
        `INSERT INTO uid_transfers (uid_id, from_location_id, to_location_id, transferred_by_id, reason)
         VALUES ($1,$2,$3,$4,$5)`,
        [uidId, uid.factory_location_id, to_location_id, req.user.id, reason]
      );
      await c.query('UPDATE uids SET factory_location_id = $1 WHERE id = $2', [to_location_id, uidId]);
    });
    res.json({ transferred: true, new_location_id: to_location_id });
  })
);

export default router;

File: backend/app/routers/uid.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session, joinedload
from sqlalchemy import or_
from pydantic import BaseModel
from typing import Optional, List
from app.database import get_db
from app.models.uid import UID, UIDStepHistory, UIDTransfer, UIDStatus, PriorityLevel
from app.models.product import DesignSizeValidity
from app.auth import require_manager, require_supervisor, require_operator, get_current_user
from app.services.uid_service import bulk_create_uids, complete_step, do_converting

router = APIRouter(prefix="/api/uids", tags=["uids"])


def uid_out(u: UID, include_history: bool = False) -> dict:
    data = {
        "id": u.id,
        "code": u.code,
        "status": u.status,
        "priority": u.priority,
        "factory_location_id": u.factory_location_id,
        "factory_location_code": u.factory_location.code if u.factory_location else None,
        "cycle_type_id": u.cycle_type_id,
        "cycle_type_name": u.cycle_type.name if u.cycle_type else None,
        "cycle_version_id": u.cycle_version_id,
        "current_step_id": u.current_step_id,
        "current_step_number": u.current_step.step_number if u.current_step else None,
        "current_step_name": u.current_step.operation_name if u.current_step else None,
        "current_storage_id": u.current_storage_id,
        "current_storage_code": u.current_storage.code if u.current_storage else None,
        "product_type_id": u.product_type_id,
        "size_id": u.size_id,
        "size_mm": u.size.value_mm if u.size else None,
        "design_id": u.design_id,
        "design_code": u.design.code if u.design else None,
        "design_confirmed": u.design_confirmed,
        "design_locked": u.design_locked,
        "mo_id": u.mo_id,
        "mo_number": u.mo.mo_number if u.mo else None,
        "parent_uid_id": u.parent_uid_id,
        "parent_uid_code": u.parent_uid.code if u.parent_uid else None,
        "child_suffix": u.child_suffix,
        "children": [{"id": c.id, "code": c.code, "status": c.status} for c in (u.children or [])],
        "created_at": u.created_at,
        "notes": u.notes,
        # Material traceability
        "faridabad_dispatch_id": u.faridabad_dispatch_id,
        "receiving_event_id": u.receiving_event_id,
        "alloy_supplier": u.alloy_supplier,
        "alloy_grade": u.alloy_grade,
        "alloy_heat_number": u.alloy_heat_number,
        "ms_supplier": u.ms_supplier,
        "ms_grade": u.ms_grade,
        "ms_heat_number": u.ms_heat_number,
        "rolling_contractor": u.rolling_contractor,
    }
    if include_history:
        data["step_history"] = [
            {
                "id": h.id,
                "cycle_step_id": h.cycle_step_id,
                "step_number": h.cycle_step.step_number if h.cycle_step else None,
                "operation_name": h.cycle_step.operation_name if h.cycle_step else None,
                "workstation_code": h.workstation.code if h.workstation else None,
                "performed_by": h.performed_by_user.full_name if h.performed_by_user else None,
                "performed_at": h.performed_at,
                "qc_result": h.qc_result,
                "qc_values": h.qc_values,
                "notes": h.notes,
                "child_uids_created": h.child_uids_created,
            }
            for h in u.step_history
        ]
    return data


# ── Lookup (service team + all) ────────────────────────────────────────────────

@router.get("/lookup/{code}")
def lookup_uid(code: str, db: Session = Depends(get_db), _=Depends(get_current_user)):
    u = db.query(UID).filter(UID.code == code.upper()).first()
    if not u:
        raise HTTPException(status_code=404, detail="UID not found")
    return uid_out(u, include_history=True)


# ── List / search UIDs ────────────────────────────────────────────────────────

@router.get("/")
def list_uids(
    location_id: Optional[int] = None,
    cycle_type_id: Optional[int] = None,
    status: Optional[UIDStatus] = None,
    step_number: Optional[str] = None,
    search: Optional[str] = None,
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
    _=Depends(get_current_user),
):
    q = db.query(UID)
    if location_id:
        q = q.filter(UID.factory_location_id == location_id)
    if cycle_type_id:
        q = q.filter(UID.cycle_type_id == cycle_type_id)
    if status:
        q = q.filter(UID.status == status)
    if search:
        q = q.filter(UID.code.ilike(f"%{search}%"))
    total = q.count()
    uids = q.order_by(UID.id.desc()).offset(skip).limit(limit).all()
    return {"total": total, "items": [uid_out(u) for u in uids]}


# ── Operator queue: UIDs at current user's workstation step ───────────────────

@router.get("/queue/operator")
def operator_queue(
    location_id: Optional[int] = None,
    db: Session = Depends(get_db),
    user=Depends(get_current_user),
):
    loc_id = location_id or user.primary_location_id
    q = db.query(UID).filter(UID.status == UIDStatus.active)
    if loc_id:
        q = q.filter(UID.factory_location_id == loc_id)
    return [uid_out(u) for u in q.order_by(UID.priority.desc(), UID.created_at).limit(200).all()]


# ── Bulk create ────────────────────────────────────────────────────────────────

class BulkCreateRequest(BaseModel):
    quantity: int
    cycle_type_id: int
    factory_location_id: int
    product_type_id: Optional[int] = None
    size_id: Optional[int] = None
    design_id: Optional[int] = None
    priority: PriorityLevel = PriorityLevel.normal
    mo_id: Optional[int] = None


@router.post("/bulk-create", status_code=201)
def bulk_create(body: BulkCreateRequest, db: Session = Depends(get_db), user=Depends(require_manager)):
    if body.quantity < 1 or body.quantity > 500:
        raise HTTPException(status_code=400, detail="Quantity must be 1–500")
    uids = bulk_create_uids(
        db,
        quantity=body.quantity,
        cycle_type_id=body.cycle_type_id,
        factory_location_id=body.factory_location_id,
        created_by_id=user.id,
        product_type_id=body.product_type_id,
        size_id=body.size_id,
        design_id=body.design_id,
        priority=body.priority,
        mo_id=body.mo_id,
    )
    return {"created": len(uids), "uids": [{"id": u.id, "code": u.code} for u in uids]}


# ── Step completion ────────────────────────────────────────────────────────────

class StepCompleteRequest(BaseModel):
    workstation_id: int
    qc_result: Optional[str] = None
    qc_values: Optional[dict] = None
    notes: Optional[str] = None


@router.post("/{uid_id}/complete-step")
def complete_uid_step(uid_id: int, body: StepCompleteRequest, db: Session = Depends(get_db), user=Depends(require_operator)):
    uid = complete_step(db, uid_id, user.id, body.workstation_id, body.qc_result, body.qc_values, body.notes)
    return uid_out(uid)


# ── Converting ────────────────────────────────────────────────────────────────

class ChildData(BaseModel):
    length_mm: Optional[int] = None
    cycle_type_id: int


class ConvertRequest(BaseModel):
    children: List[ChildData]
    pattern_id: Optional[int] = None


@router.post("/{uid_id}/convert")
def convert_uid(uid_id: int, body: ConvertRequest, db: Session = Depends(get_db), user=Depends(require_supervisor)):
    children = do_converting(
        db, uid_id, user.id,
        [c.model_dump() for c in body.children],
        body.pattern_id,
    )
    return {"parent_uid_id": uid_id, "children": [uid_out(c) for c in children]}


# ── Design confirmation ───────────────────────────────────────────────────────

class DesignConfirmRequest(BaseModel):
    design_id: int
    size_id: Optional[int] = None


@router.post("/{uid_id}/confirm-design")
def confirm_design(uid_id: int, body: DesignConfirmRequest, db: Session = Depends(get_db), user=Depends(require_manager)):
    uid = db.query(UID).filter(UID.id == uid_id).first()
    if not uid:
        raise HTTPException(status_code=404, detail="UID not found")
    if uid.design_locked:
        raise HTTPException(status_code=400, detail="Design is locked — cannot change after Step 17")

    # Validate design-size combination
    if body.size_id:
        valid = db.query(DesignSizeValidity).filter(
            DesignSizeValidity.design_id == body.design_id,
            DesignSizeValidity.size_id == body.size_id,
        ).first()
        if not valid:
            raise HTTPException(status_code=400, detail="Invalid design-size combination")
        uid.size_id = body.size_id

    uid.design_id = body.design_id
    uid.design_confirmed = True
    if uid.status == UIDStatus.on_hold:
        uid.status = UIDStatus.active
    db.commit()
    db.refresh(uid)
    return uid_out(uid)


# ── MO Linking ────────────────────────────────────────────────────────────────

@router.post("/{uid_id}/link-mo/{mo_id}")
def link_mo(uid_id: int, mo_id: int, db: Session = Depends(get_db), user=Depends(require_manager)):
    uid = db.query(UID).filter(UID.id == uid_id).first()
    if not uid:
        raise HTTPException(status_code=404, detail="UID not found")
    uid.mo_id = mo_id
    db.commit()
    return {"uid_id": uid_id, "mo_id": mo_id}


# ── Inter-location transfer ────────────────────────────────────────────────────

class TransferRequest(BaseModel):
    to_location_id: int
    reason: str


@router.post("/{uid_id}/transfer")
def transfer_uid(uid_id: int, body: TransferRequest, db: Session = Depends(get_db), user=Depends(require_supervisor)):
    uid = db.query(UID).filter(UID.id == uid_id).first()
    if not uid:
        raise HTTPException(status_code=404, detail="UID not found")
    transfer = UIDTransfer(
        uid_id=uid_id,
        from_location_id=uid.factory_location_id,
        to_location_id=body.to_location_id,
        transferred_by_id=user.id,
        reason=body.reason,
    )
    uid.factory_location_id = body.to_location_id
    db.add(transfer)
    db.commit()
    return {"transferred": True, "new_location_id": body.to_location_id}


# ── Bulk cycle type change (before any steps) ──────────────────────────────────

class BulkCycleChange(BaseModel):
    uid_ids: List[int]
    new_cycle_type_id: int


@router.post("/bulk-change-cycle")
def bulk_change_cycle(body: BulkCycleChange, db: Session = Depends(get_db), user=Depends(require_manager)):
    from app.models.cycle import CycleVersion
    new_version = db.query(CycleVersion).filter(
        CycleVersion.cycle_type_id == body.new_cycle_type_id,
        CycleVersion.is_current == True,
    ).first()
    if not new_version:
        raise HTTPException(status_code=404, detail="Cycle type or version not found")

    updated = []
    for uid_id in body.uid_ids:
        uid = db.query(UID).filter(UID.id == uid_id).first()
        if not uid:
            continue
        if uid.step_history:
            raise HTTPException(status_code=400, detail=f"UID {uid.code} has steps completed — cannot change cycle")
        uid.cycle_type_id = body.new_cycle_type_id
        uid.cycle_version_id = new_version.id
        updated.append(uid.code)

    db.commit()
    return {"updated": updated}
